### PR TITLE
fix: Workaround `postconf` write settling logic

### LINKED ISSUE #2985: [TODO] Remove unnecessary 2 second delays from Postfix utilities during tests
## Description

This issue is mostly to document an observation that impacts testing time from some Postfix utilities introducing 2 second delays when used if `/etc/postfix/main.cf` has been modified within that time frame.

---

### Background

In May 2022 a change was made for how Postfix config overrides were applied. An observation was noticed [here](https://github.com/docker-mailserver/docker-mailserver/pull/2598#issuecomment-1141561786). The [original approach](https://github.com/docker-mailserver/docker-mailserver/pull/2598/files#diff-141cf3167e0d3e34aa4f5242772a4e7519a0abb8a7cd2ba4493e8ea0ed607a61L738-L747) does not incur the observed delay.

Recently I investigated this some more to reduce a few seconds from every container spun up for CI tests. I had noticed that if the configs for that override logic weren't present, 2 seconds was shaved off. And another part of the startup scripts was adding 2 seconds as well with `postalias /etc/aliases`. It turns out these Postfix utilities are reading `/etc/postfix/main.cf` during execution.

I found `pausing to let /etc/postfix/main.cf cool down` would be output with `-vv` verbosity set when running commands. This prompted to look into the source for Postfix utilities, I found the [following polling / delay logic](https://github.com/vdukhovni/postfix/blob/1f2df46e5211065a16f78739308131122e185c7c/postfix/src/util/load_file.c#L76-L78) (_`doze()` is another utility method in the source that takes microseconds_). It compares the `mtime` of the file to load to be less than 1 second passed or a value in the future, and waits 300ms until checking again.

Thus anytime startup modifies `/etc/postfix/main.cf` and tries to run one of the postfix utilities shortly after, there is a delay measured of about 1.5-2 seconds. 

`postconf 'max_idle = 600s'` can be run with no penalty. But `postconf max_idle` to perform a lookup will incur that delay.

---

## To consider

This is probably only a concern with tests where the start-up time can be compounded for common start-up logic.

- Only one test actually hit the config override path, and that has been [addressed in a recent PR](https://github.com/docker-mailserver/docker-mailserver/pull/2980/files#diff-1420e606ddac9d72de297d6a257e6c98fbf68c4dd7a5ee6f985d27f7ec09edda).
- All tests presently add a default `root` postmaster alias. This hit can be worked around by modifying the `mtime` (_`stat -c %Y /path/to/file` for checking this value as unix epoch timestamp_) via `touch -d '2 seconds ago' /path/to/file`. I think that should be fine? (_This also reduces the output spam from the original config override observation PR, avoids the repetition_)
- Other tests or logic may need to be evaluated for similar delays incurred by this stalling behaviour (_although AFAIU it's intended to ensure a file is actually ready to be read by no longer being written to_). An alternative approach may be to replace the `sed` + `postconf` calls and instead queue up the modifications to be applied.

### PR DESCRIPTION
**Pull request by polarathene**
# Description

After updating `main.cf`, to avoid an enforced delay from reading the config by postfix tools, we can ensure the modified time is at least 2 seconds in the past as a workaround. This should be ok with our usage AFAIK.

Shaves off 2+ seconds roughly off each container startup, reduces roughly 2+ minutes off tests.

Fixes #2985

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by georglauterbach — DISMISSED** [review]
LGTM 👍🏼 Nice work!

**Comment by georglauterbach** [comment]
The pipeline `Test Merge Requests / Test AMD64 Image / Test (serial) (pull_request)` is now `Successful in 13m`. We're already pretty close to 12min with the tests. When we improve on this further, and we definitely can (and will :D), we will see times lower than 12mins, so the next logical step when it comes to improving the CI is to speed up the ARM64 build. @polarathene @casperklein @wernerfred do you have any ideas how to do that?

**Comment by polarathene** [comment]
> When we improve on this further, and we definitely can (and will :D), we will see times lower than 12mins

Yes definitely possible. There's still the supervisorctl process restarting checks that can all be moved into a single test file I think. Presently I am just focusing on adjusting existing serial tests to the new format, and then they'll be distributed into the parallel sets which will bring the time down nicely :+1: 

> do you have any ideas how to do that?

It would require leveraging the image cache well. It's a bit tricky with Github CI cache keys IIRC. On a new PR, it can use a cache key from master branch, but master cannot use one from a PR, nor can other PRs use  cache keys created by adjacent PRs.

I'd have to look over it some time to refresh what can be done. I do recall there is an experimental cache mode for Github CI, our buildx action has a `gha` mode that's meant to be better at caching IIRC. I won't be investigating that until I've tackled converting all these tests. Even then I've still got other tasks to prioritize :sweat_smile:

**Comment by polarathene** [comment]
> I will be doing that as well to help you in this regard.

I don't mind raising the PRs, I was slowed down on the `mail_changedetector` one to remove the sleeps as it required some extra work on the change detection helper and understanding the test better, extra careful not to mess it up :sweat_smile: 

If you are planning to update other tests do gives a heads-up on which ones you're going to tackle. I would also encourage just converting them without the rename, so we can just do a PR afterwards that handles that for preserving history :+1: 

I plan to spend some time on `mail_hostname` (_if I can find my old local branch that had a revision of it sitting around from a while back :grimacing: )

**Comment by georglauterbach** [comment]
> If you are planning to update other tests do gives a heads-up on which ones you're going to tackle. I would also encourage just converting them without the rename, so we can just do a PR afterwards that handles that for preserving history 👍

I will be tackling

- `mail_time`
- `sedfile`
- `no_container`

next :D

> I plan to spend some time on `mail_hostname` (_if I can find my old local branch that had a revision of it sitting around from a while back 😬 )

There are plans to revise the DNS name stuff (I already added the [new ENV variables to the scripts](https://github.com/docker-mailserver/docker-mailserver/blob/master/target/scripts/helpers/variables.sh#L104-L107)). Probably worthwhile investigating this first and adjusting the tests according to these changes?

**Review by casperklein — APPROVED** [review]

**Comment by polarathene** [comment]
> I'm thinking perhaps to add a check on the `mtime` being `< 2 seconds ago` into this util method. Do you think it'd be useful?

- I've now added the conditional check for this, and added it into `check-for-changes.sh` which may benefit from it.
- Also identified a few other locations where it should be applied. Since it was needed across a few tests that manually modify `main.cf`, I've added a helper method into `common.bash`.

**Comment by polarathene** [comment]
[Failure](https://github.com/docker-mailserver/docker-mailserver/actions/runs/3897558295/jobs/6655376329#step:5:64):

```
not ok 56 checking smtp_only: mail send should work in 1327ms
# (from function `_reload_postfix' in file test/test_helper/common.bash, line 17,
#  in test file test/tests/serial/mail_smtponly.bats, line 49)
#   `_reload_postfix mail_smtponly' failed
```

Reproduced locally, looking into it :sweat_smile:

**Comment by polarathene** [comment]
Identified that:

- `docker exec "${CONTAINER_NAME}" sh -c "/etc/init.d/postfix reload"` works.
- `docker exec "${CONTAINER_NAME}" sh -c "service postfix reload"` works.
- `docker exec "${CONTAINER_NAME}" sh -c "postfix reload"` **fails**.

Yet it only fails in this test :thinking: 

---

I've got a separate branch where I was looking into removing the need for the `postfix-wrapper.sh`, and when I apply the changes for that, then `sh -c 'postfix reload'` does not fail. So it's presumably something to do with `postfix-wrapper.sh`?

Maybe this PR should be deferred until I have that related change (remove needing `postfix-wrapper.sh`) considered for merging?

**Review by georglauterbach — DISMISSED** [review]

File: target/bin/restrict-access [hunk 1]
```diff
@@ -44,7 +44,8 @@ case "${MODE}" in
           sed -i 's|smtpd_sender_restrictions =|smtpd_sender_restrictions = check_sender_access texthash:/tmp/docker-mailserver/postfix-send-access.cf,|' /etc/postfix/main.cf \
         || sed -i 's|smtpd_recipient_restrictions =|smtpd_recipient_restrictions = check_recipient_access texthash:/tmp/docker-mailserver/postfix-receive-access.cf,|' /etc/postfix/main.cf
 
-      service postfix reload >/dev/null
+      _adjust_mtime_for_postfix_maincf
+      postfix reload
```
Comment (casperklein) [inline target/bin/restrict-access:48]:
```suggestion
      _reload_postfix
```

if `_reload_postfix` function is moved from `check-for-changes.sh` to `utils.sh`, the function can be used here.

**Comment by casperklein** [comment]
> * I've now added the conditional check for this

Just to be save or do you expect any problems?

**Comment by polarathene** [comment]
> Just to be save or do you expect any problems?

Not from what I understand. In the [associated issue](https://github.com/docker-mailserver/docker-mailserver/issues/2985) I linked to the source code upstream in Postfix that is responsible for the delay.

It's a safety guard for reading `main.cf` by checking `mtime` isn't too recent so that it can be assumed safe to read (_not in a partially updated state, all writes completed / settled_). It's a valid condition to wait on to ensure correct config is read.

I believe it's safe to read after make changes, but perhaps depending on hardware or third-party software something may not play well (or try to abuse/exploit it?). Postfix will tend to ignore anything that it doesn't recognize as valid config keys, which could result in a default setting used instead, or the key is valid but the value is not the intended value? I think exploiting sensitive timing like that is a difficult attack and if that were an option the system is likely already vulnerable to much worse?

---

It's mostly tests that will benefit from this, or in actual deployments minimal downtime during change events (_[seems to be an actual concern](https://github.com/docker-mailserver/docker-mailserver/issues/2613), I think there are some similar issues reported_). Additionally reduces container start-up time, but I doubt that matters much to most users.

I am inclined to say it should be fine to workaround. It's usually more of an issue when wanting to write to a file that other processes might also attempt, or reading while a file is actively being written to (what this protects against).

**Review by georglauterbach — APPROVED** [review]

**Review by casperklein — APPROVED** [review]

**Pull request merged**